gcc/rust/expand/rust-macro-builtins-asm.cc

```diff
@@ -16,5 +16,528 @@
+
+  // after successful left parenthesis parsing, we should return ast of
+  // InlineAsmRegOrRegClass of reg or reg class
+  auto token = parser.peek_current_token ();
+  auto tok_id = token->get_id ();
+  AST::InlineAsmRegOrRegClass regClass;
+  if (parser.skip_token (IDENTIFIER))
+    {
+      // construct a InlineAsmRegOrRegClass
+      regClass.type = RegType::RegClass;
+      regClass.regClass.Symbol = token->as_string ();
+    }
+  else if (tok_id == STRING_LITERAL)
+    {
+      // TODO: there is STRING_LITERAL, and BYTE_STRING_LITERAL, should we check
+      // for both?
+
+      // construct a InlineAsmRegOrRegClass
+      // parse_format_string
+      regClass.type = RegType::Reg;
+      inlineAsmCtx.is_explicit = true;
+      regClass.regClass.Symbol = token->as_string ();
+    }
+  else
+    {
+      // TODO: This should emit error
+      //  return
+      //  Err(p.dcx().create_err(errors::ExpectedRegisterClassOrExplicitRegister
+      //  {
+      //           span: p.token.span,
+      //       }));
+    }
+  if (!parser.skip_token (RIGHT_PAREN))
+    {
+      // TODO: we expect a left parenthesis here, please return the correct
+      // error.
+      return tl::nullopt;
+    }
+
+  return regClass;
+}
+
+int
+parse_operand (Parser<MacroInvocLexer> &parser, TokenId last_token_id,
+	       InlineAsmContext &inlineAsmCtx)
+{
+  return 0;
+}
+
+// From rustc
+tl::optional<AST::InlineAsmOperand>
+parse_reg_operand (Parser<MacroInvocLexer> &parser, TokenId last_token_id,
+		   InlineAsmContext &inlineAsmCtx)
+{
+  // let name = if p.token.is_ident() && p.look_ahead(1, |t| *t == token::Eq) {
+  //           let (ident, _) = p.token.ident().unwrap();
+  //           p.bump();
+  //           p.expect(&token::Eq)?;
+  //           allow_templates = false;
+  //           Some(ident.name)
+  //       } else {
+  //           None
+  //       };
+
+  using RegisterType = AST::InlineAsmOperand::RegisterType;
+  AST::InlineAsmOperand reg_operand;
+  auto token = parser.peek_current_token ();
+  auto iden_token = parser.peek_current_token ();
+  auto &inlineAsm = inlineAsmCtx.inlineAsm;
+  if (check_identifier (parser, ""))
+    {
+      auto equal_token = parser.peek_current_token ();
+      if (!parser.skip_token (EQUAL))
+	{
+	  // TODO: Please handle this.
+	  rust_unreachable ();
+	}
+    }
+
+  token = parser.peek_current_token ();
+
+  bool is_explicit_reg = false;
+  bool is_global_asm = inlineAsm.is_global_asm;
+  if (!is_global_asm && check_identifier (parser, "in"))
+    {
+      return tl::nullopt;
+    }
+  else if (!is_global_asm && check_identifier (parser, "out"))
+    {
+      return tl::nullopt;
+    }
+  else if (!is_global_asm && check_identifier (parser, "lateout"))
+    {
+      return tl::nullopt;
+    }
+  else if (!is_global_asm && check_identifier (parser, "inout"))
+    {
+      return tl::nullopt;
+    }
+  else if (!is_global_asm && check_identifier (parser, "inlateout"))
+    {
+      return tl::nullopt;
+    }
```

Comment:
I think this might be cleaner if it was turned into an array of strings to check for, and if any matches, return `tl::nullopt`. what do we check here exactly?

Reply:
this is will eventually constructor instances of RegisterOperand class and here i'm just setting up the scaffolding to match rust's asm.rs for easier matching right now. Different lines of code should come out from these if statements

Reply:
we're checking if the identifier matches each of the name, which stands for `dir_spec := "in" / "out" / "lateout" / "inout" / "inlateout"` in the ABNF. i guess it also goes with whether the asm call is global or not too so we also need to check for that